File: frontend/lib/services/loan_service.dart
import 'package:flutter/foundation.dart';

import '../models/emi.dart';
import '../models/enums.dart';
import '../models/loan.dart';
import '../utils/id_gen.dart';
import 'gate.dart';

abstract class LoanService extends ChangeNotifier {
  List<Loan> all();
  Loan? byId(String id);
  List<Loan> forCustomer(String customerId);

  /// Books a loan against a customer (and optionally a vehicle). No interest:
  /// each month owes the typed [emiAmount], repeated for [tenureMonths].
  Loan create({
    required Role actorRole,
    required String actorId,
    required String customerId,
    String? vehicleId,
    required int principal,
    required int tenureMonths,
    required int emiAmount,
    required DateTime disbursementDate,
  });

  /// Records a (possibly partial) payment against one EMI. [penalty] is the late
  /// penalty for the month, [amount] is what is being paid now; [receivedDate],
  /// [remarks] and [screenshotName] capture the proof/notes. Auto-closes the
  /// loan once every EMI is fully cleared.
  void recordEmiPayment(
    String loanId,
    String emiId,
    int amount, {
    int penalty = 0,
    DateTime? receivedDate,
    String? remarks,
    String? screenshotName,
    Uint8List? screenshotBytes,
    String? screenshotMime,
  });

  /// Full edit within the 5-hour grace window: replaces every detail and
  /// REBUILDS the EMI schedule, discarding any recorded payments (the caller
  /// warns first). The server rejects it once the window has passed.
  void edit(
    String loanId, {
    required String customerId,
    String? vehicleId,
    required int principal,
    required int tenureMonths,
    required int emiAmount,
    required DateTime disbursementDate,
    String? remarks,
  });

  void foreclose(String loanId, {int charge});
  void waivePenalty(String loanId, String emiId);
  void confirm(String id, String byUserId);
  void reject(String id, String reason, String byUserId);
  void delete(String id);

  // ── Seizure (repossession) ─────────────────────────────────────────────────
  /// Requests seizing the loan vehicle. A super admin's request seizes at once;
  /// an admin's waits pending a super admin's confirm/cancel.
  void requestSeize(String loanId, String reason,
      {required bool superAdmin, required String byUserId});
  void confirmSeize(String loanId, String byUserId);

  /// Cancels a seize — the vehicle goes back to the customer, loan continues.
  void cancelSeize(String loanId, String byUserId, {String? remarks});

  // ── Payment screenshots ─────────────────────────────────────────────────────
  /// Raw bytes of a stored payment screenshot, for in-app preview.
  Future<Uint8List> paymentDocBytes(int docId);

  /// Absolute URL of a stored payment screenshot.
  String paymentDocUrl(int docId);

  /// Deletes a stored payment screenshot by its backend id.
  void deletePaymentDoc(String loanId, int docId);
}

class MockLoanService extends LoanService {
  MockLoanService() {
    _seed();
  }

  final List<Loan> _loans = [];

  /// Builds a flat, no-interest EMI schedule: [tenureMonths] rows each owing
  /// [emiAmount], due one month apart starting at [firstDue].
  static List<Emi> buildSchedule({
    required int emiAmount,
    required int tenureMonths,
    required DateTime firstDue,
    int paidUpTo = 0,
  }) {
    return List.generate(tenureMonths, (i) {
      final due = DateTime(firstDue.year, firstDue.month + i, firstDue.day);
      return Emi(
        id: IdGen.nextId('emi'),
        sequenceNumber: i + 1,
        dueDate: due,
        amountDue: emiAmount,
        amountPaid: i < paidUpTo ? emiAmount : 0,
        paidDate: i < paidUpTo ? due : null,
        receivedDate: i < paidUpTo ? due : null,
      );
    });
  }

  void _seed() {
    // Active loan, 2 of 20 EMIs paid — ₹7,000/month.
    _loans.add(Loan(
      id: IdGen.nextId('loan'),
      customerId: 'c_001',
      principal: 140000,
      tenureMonths: 20,
      disbursementDate: DateTime(2026, 6, 5),
      firstEmiDueDate: DateTime(2026, 7, 5),
      emiAmount: 7000,
      emis: buildSchedule(
        emiAmount: 7000,
        tenureMonths: 20,
        firstDue: DateTime(2026, 7, 5),
        paidUpTo: 1,
      ),
      loanStatus: 'active',
      createdBy: 'u_super',
      createdAt: DateTime(2026, 6, 5),
      status: EntityStatus.active,
    ));
  }

  @override
  List<Loan> all() => List.unmodifiable(_loans);

  @override
  Loan? byId(String id) =>
      _loans.where((l) => l.id == id).cast<Loan?>().firstOrNull;

  @override
  List<Loan> forCustomer(String customerId) =>
      _loans.where((l) => l.customerId == customerId).toList();

  @override
  Loan create({
    required Role actorRole,
    required String actorId,
    required String customerId,
    String? vehicleId,
    required int principal,
    required int tenureMonths,
    required int emiAmount,
    required DateTime disbursementDate,
  }) {
    final firstDue = DateTime(
        disbursementDate.year, disbursementDate.month + 1, disbursementDate.day);
    final loan = Loan(
      id: IdGen.nextId('loan'),
      customerId: customerId,
      vehicleId: vehicleId,
      principal: principal,
      tenureMonths: tenureMonths,
      disbursementDate: disbursementDate,
      firstEmiDueDate: firstDue,
      emiAmount: emiAmount,
      emis: buildSchedule(
        emiAmount: emiAmount,
        tenureMonths: tenureMonths,
        firstDue: firstDue,
      ),
      loanStatus: 'active',
      createdBy: actorId,
      createdAt: DateTime.now(),
      status: Gate.initialStatus(actorRole),
    );
    _loans.insert(0, loan);
    notifyListeners();
    return loan;
  }

  @override
  void recordEmiPayment(
    String loanId,
    String emiId,
    int amount, {
    int penalty = 0,
    DateTime? receivedDate,
    String? remarks,
    String? screenshotName,
    Uint8List? screenshotBytes,
    String? screenshotMime,
  }) {
    final loan = byId(loanId);
    final emi = loan?.emis.where((e) => e.id == emiId).cast<Emi?>().firstOrNull;
    if (loan == null || emi == null) return;

    emi.penalty = penalty;
    emi.amountPaid = (emi.amountPaid + amount).clamp(0, emi.totalDue);
    emi.receivedDate = receivedDate ?? DateTime.now();
    if (remarks != null && remarks.isNotEmpty) emi.remarks = remarks;
    if (screenshotName != null && screenshotName.isNotEmpty) {
      emi.screenshotName = screenshotName;
    }
    if (emi.isPaid) emi.paidDate = emi.receivedDate;

    if (loan.emis.every((e) => e.isPaid)) {
      loan.loanStatus = 'closed';
    } else if (loan.emis.any((e) =>
        !e.isPaid && e.statusAt(DateTime.now()) == ScheduleStatus.overdue)) {
      loan.loanStatus = 'overdue';
    } else {
      loan.loanStatus = 'active';
    }
    notifyListeners();
  }

  @override
  void edit(
    String loanId, {
    required String customerId,
    String? vehicleId,
    required int principal,
    required int tenureMonths,
    required int emiAmount,
    required DateTime disbursementDate,
    String? remarks,
  }) {
    final i = _loans.indexWhere((l) => l.id == loanId);
    if (i < 0) return;
    final old = _loans[i];
    final firstDue = DateTime(
        disbursementDate.year, disbursementDate.month + 1, disbursementDate.day);
    _loans[i] = Loan(
      id: old.id,
      customerId: customerId,
      vehicleId: vehicleId,
      principal: principal,
      tenureMonths: tenureMonths,
      disbursementDate: disbursementDate,
      firstEmiDueDate: firstDue,
      emiAmount: emiAmount,
      emis: buildSchedule(
        emiAmount: emiAmount, tenureMonths: tenureMonths, firstDue: firstDue),
      loanStatus: 'active',
      createdBy: old.createdBy,
      createdAt: old.createdAt,
      status: old.status,
      confirmedBy: old.confirmedBy,
      confirmedAt: old.confirmedAt,
    );
    notifyListeners();
  }

  @override
  void foreclose(String loanId, {int charge = 0}) {
    final loan = byId(loanId);
    if (loan == null) return;
    for (final e in loan.emis) {
      if (!e.isPaid) {
        e.amountPaid = e.totalDue;
        e.paidDate = DateTime.now();
        e.receivedDate = DateTime.now();
      }
    }
    loan.loanStatus = 'foreclosed';
    notifyListeners();
  }

  @override
  void waivePenalty(String loanId, String emiId) {
    final loan = byId(loanId);
    final emi =
        loan?.emis.where((e) => e.id == emiId).cast<Emi?>().firstOrNull;
    if (emi != null) {
      emi.penalty = 0;
      notifyListeners();
    }
  }

  @override
  void confirm(String id, String byUserId) {
    final l = byId(id);
    if (l != null) {
      Gate.confirm(l, byUserId: byUserId);
      notifyListeners();
    }
  }

  @override
  void reject(String id, String reason, String byUserId) {
    final l = byId(id);
    if (l != null) {
      Gate.reject(l, reason: reason, byUserId: byUserId);
      l.loanStatus = 'rejected';
      notifyListeners();
    }
  }

  @override
  void delete(String id) {
    _loans.removeWhere((l) => l.id == id);
    notifyListeners();
  }

  @override
  void requestSeize(String loanId, String reason,
      {required bool superAdmin, required String byUserId}) {
    final l = byId(loanId);
    if (l == null) return;
    l.seizeReason = reason;
    l.seizedAt = DateTime.now();
    if (superAdmin) {
      l.seizeStage = 'seized';
      l.loanStatus = 'seized';
    } else {
      l.seizeStage = 'pending';
    }
    notifyListeners();
  }

  @override
  void confirmSeize(String loanId, String byUserId) {
    final l = byId(loanId);
    if (l == null) return;
    l.seizeStage = 'seized';
    l.loanStatus = 'seized';
    notifyListeners();
  }

  @override
  void cancelSeize(String loanId, String byUserId, {String? remarks}) {
    final l = byId(loanId);
    if (l == null) return;
    l.seizeStage = null;
    if (l.loanStatus == 'seized') l.loanStatus = 'active';
    notifyListeners();
  }

  @override
  Future<Uint8List> paymentDocBytes(int docId) async => Uint8List(0);

  @override
  String paymentDocUrl(int docId) => '';

  @override
  void deletePaymentDoc(String loanId, int docId) {}
}

File: backend/app/services/loan_service.py
"""Business logic for loans — no interest, typed EMI, penalty on late months.

Role-gate mirrors the rest of the app: a Super Admin's loan is active at once;
an Admin's loan waits pending_confirmation until a Super Admin approves it.
"""
import calendar
from datetime import date, datetime, timedelta, timezone
from decimal import Decimal

from fastapi import HTTPException
from sqlalchemy.orm import Session

from app.dao.loan_dao import LoanDAO
from app.models.enums import EntityStatus, InstallmentStatus, NotificationEntity
from app.models.loan import Loan
from app.models.loan_emi import LoanEmi
from app.models.loan_payment import LoanPayment
from app.models.loan_payment_document import LoanPaymentDocument
from app.schemas.loan import LoanCreate, LoanEdit
from app.services.notification_service import NotificationService
from app.services.vehicle_service import initial_status

# How long after booking a loan stays editable. After this the schedule is
# locked and the client hides the Edit button (the server enforces it too).
EDIT_WINDOW = timedelta(hours=5)


def _add_months(d: date, n: int) -> date:
    """Date n months after d, clamping the day to the target month's length."""
    m = d.month - 1 + n
    y = d.year + m // 12
    m = m % 12 + 1
    last = calendar.monthrange(y, m)[1]
    return date(y, m, min(d.day, last))


class LoanService:
    @staticmethod
    def list(db: Session, *, module=None, status=None, customer_id=None) -> list[Loan]:
        return LoanDAO.list(db, module=module, status=status, customer_id=customer_id)

    @staticmethod
    def get(db: Session, loan_id: int) -> Loan:
        loan = LoanDAO.get(db, loan_id)
        if loan is None:
            raise HTTPException(status_code=404, detail="Loan not found")
        return loan

    @staticmethod
    def create(
        db: Session, data: LoanCreate, *, actor_role: str, created_by: int
    ) -> Loan:
        module_id = LoanDAO.module_id_by_code(db, data.module_code)
        if module_id is None:
            raise HTTPException(status_code=400, detail=f"Unknown module '{data.module_code}'")
        if data.tenure_months <= 0:
            raise HTTPException(status_code=400, detail="Tenure must be at least 1 month")

        status = data.status or initial_status(actor_role)
        first_due = _add_months(data.loan_date, 1)
        loan = Loan(
            module_id=module_id,
            customer_id=data.customer_id,
            vehicle_id=data.vehicle_id,
            principal=data.principal,
            emi_amount=data.emi_amount,
            tenure_months=data.tenure_months,
            loan_date=data.loan_date,
            first_due_date=first_due,
            loan_status="active",
            status=status,
            created_by=created_by,
            remarks=data.remarks,
        )
        # Build the flat, no-interest EMI schedule: one row per month.
        for i in range(data.tenure_months):
            loan.emis.append(
                LoanEmi(
                    module_id=module_id,
                    sequence_number=i + 1,
                    due_date=_add_months(data.loan_date, i + 1),
                    amount=data.emi_amount,
                    status=InstallmentStatus.pending,
                )
            )
        LoanDAO.add(db, loan)

        if actor_role != "super_admin":
            NotificationService.create_verification(
                db,
                entity_type=NotificationEntity.loan,
                entity_id=loan.id,
                title="New loan needs approval",
                message="A loan created by an admin awaits verification.",
            )
        db.commit()
        return LoanDAO.get(db, loan.id)

    @staticmethod
    def _created_at_utc(loan: Loan) -> datetime:
        """loan.created_at as an aware UTC datetime (the column is naive)."""
        c = loan.created_at
        return c.replace(tzinfo=timezone.utc) if c.tzinfo is None else c

    @staticmethod
    def is_editable(loan: Loan, now: datetime | None = None) -> bool:
        """True while the loan is still within its post-booking edit window."""
        now = now or datetime.now(timezone.utc)
        return now - LoanService._created_at_utc(loan) <= EDIT_WINDOW

    @staticmethod
    def edit(
        db: Session, loan_id: int, data: LoanEdit, *, actor_role: str, actor_id: int
    ) -> Loan:
        """Replace a loan's details within the 5-hour grace window and REBUILD its
        EMI schedule from scratch. Any EMIs/payments already recorded are wiped —
        the client shows a warning first. Rejected once the window has passed."""
        loan = LoanService.get(db, loan_id)
        if not LoanService.is_editable(loan):
            raise HTTPException(
                status_code=403,
                detail="This loan can no longer be edited (the 5-hour window has passed).",
            )
        if loan.loan_status == "seized" or loan.seize_stage in ("pending", "seized"):
            raise HTTPException(
                status_code=400, detail="A seized loan cannot be edited."
            )
        if data.tenure_months <= 0:
            raise HTTPException(status_code=400, detail="Tenure must be at least 1 month")

        # Wipe the old schedule + every recorded payment (cascades to documents).
        loan.payments.clear()
        loan.emis.clear()
        db.flush()

        # Apply the new details.
        loan.customer_id = data.customer_id
        loan.vehicle_id = data.vehicle_id
        loan.principal = data.principal
        loan.emi_amount = data.emi_amount
        loan.tenure_months = data.tenure_months
        loan.loan_date = data.loan_date
        loan.first_due_date = _add_months(data.loan_date, 1)
        loan.remarks = data.remarks
        loan.loan_status = "active"
        loan.closed_at = None
        loan.updated_at = datetime.now(timezone.utc)

        # Rebuild the flat, no-interest EMI schedule.
        for i in range(data.tenure_months):
            loan.emis.append(
                LoanEmi(
                    module_id=loan.module_id,
                    sequence_number=i + 1,
                    due_date=_add_months(data.loan_date, i + 1),
                    amount=data.emi_amount,
                    status=InstallmentStatus.pending,
                )
            )
        db.commit()
        return LoanDAO.get(db, loan.id)

    @staticmethod
    def confirm(db: Session, loan_id: int, by_user_id: int) -> Loan:
        loan = LoanService.get(db, loan_id)
        loan.status = EntityStatus.active
        loan.confirmed_by = by_user_id
        loan.confirmed_at = datetime.now(timezone.utc)
        loan.rejection_reason = None
        db.commit()
        return LoanDAO.get(db, loan_id)

    @staticmethod
    def reject(db: Session, loan_id: int, reason: str, by_user_id: int) -> Loan:
        loan = LoanService.get(db, loan_id)
        loan.status = EntityStatus.rejected
        loan.loan_status = "rejected"
        loan.confirmed_by = by_user_id
        loan.confirmed_at = datetime.now(timezone.utc)
        loan.rejection_reason = reason
        db.commit()
        return LoanDAO.get(db, loan_id)

    @staticmethod
    def delete(db: Session, loan_id: int) -> None:
        loan = LoanService.get(db, loan_id)
        db.delete(loan)
        db.commit()

    # ── seizure (repossession) ───────────────────────────────────────────────
    @staticmethod
    def request_seize(
        db: Session, loan_id: int, reason: str, *, actor_role: str, actor_id: int
    ) -> Loan:
        loan = LoanService.get(db, loan_id)
        if loan.loan_status in ("closed", "seized"):
            raise HTTPException(status_code=400, detail="Loan already closed or seized")
        loan.seize_reason = reason
        loan.seized_by = actor_id
        loan.seized_at = datetime.now(timezone.utc)
        if actor_role == "super_admin":
            # Super admin seizes at once.
            loan.seize_stage = "seized"
            loan.seize_confirmed_by = actor_id
            loan.seize_confirmed_at = datetime.now(timezone.utc)
            loan.loan_status = "seized"
        else:
            loan.seize_stage = "pending"
            NotificationService.create_verification(
                db,
                entity_type=NotificationEntity.loan,
                entity_id=loan.id,
                title="Loan seize needs approval",
                message=f"An admin requested to seize a loan vehicle. Reason: {reason}",
            )
        db.commit()
        return LoanDAO.get(db, loan_id)

    @staticmethod
    def confirm_seize(db: Session, loan_id: int, by_user_id: int) -> Loan:
        loan = LoanService.get(db, loan_id)
        loan.seize_stage = "seized"
        loan.seize_confirmed_by = by_user_id
        loan.seize_confirmed_at = datetime.now(timezone.utc)
        loan.loan_status = "seized"
        db.commit()
        return LoanDAO.get(db, loan_id)

    @staticmethod
    def cancel_seize(
        db: Session, loan_id: int, remarks: str | None, by_user_id: int
    ) -> Loan:
        loan = LoanService.get(db, loan_id)
        loan.seize_stage = None
        loan.seize_cancel_remarks = remarks
        loan.seize_confirmed_by = by_user_id
        loan.seize_confirmed_at = datetime.now(timezone.utc)
        # Vehicle goes back to the customer — the loan continues.
        LoanService._recompute_status(loan, date.today())
        db.commit()
        return LoanDAO.get(db, loan_id)

    # ── payments ────────────────────────────────────────────────────────────────
    @staticmethod
    def record_payment(
        db: Session,
        loan_id: int,
        emi_id: int,
        *,
        amount: Decimal,
        penalty: Decimal,
        received_date: date | None,
        remarks: str | None,
        recorded_by: int,
    ) -> LoanPayment:
        loan = LoanService.get(db, loan_id)
        emi = LoanDAO.get_emi(db, emi_id)
        if emi is None or emi.loan_id != loan.id:
            raise HTTPException(status_code=404, detail="EMI not found")

        received = received_date or date.today()
        emi.penalty = penalty or Decimal(0)
        total_due = Decimal(emi.amount) + Decimal(emi.penalty)
        new_paid = Decimal(emi.amount_paid) + Decimal(amount or 0)
        emi.amount_paid = min(new_paid, total_due)
        emi.received_date = received
        if remarks:
            emi.remarks = remarks

        if Decimal(emi.amount_paid) >= total_due:
            emi.status = InstallmentStatus.paid
            emi.paid_date = received
        elif emi.due_date < received:
            emi.status = InstallmentStatus.overdue
        else:
            emi.status = InstallmentStatus.pending

        payment = LoanPayment(
            loan_id=loan.id,
            emi_id=emi.id,
            amount=amount or 0,
            penalty=penalty or 0,
            received_date=received,
            remarks=remarks,
            recorded_by=recorded_by,
        )
        LoanDAO.add_payment(db, payment)

        LoanService._recompute_status(loan, received)
        db.commit()
        db.refresh(payment)
        return payment

    @staticmethod
    def _recompute_status(loan: Loan, today: date) -> None:
        if all(e.status == InstallmentStatus.paid for e in loan.emis):
            loan.loan_status = "closed"
            loan.closed_at = datetime.now(timezone.utc)
        elif any(
            e.status != InstallmentStatus.paid and e.due_date < today for e in loan.emis
        ):
            loan.loan_status = "overdue"
        else:
            loan.loan_status = "active"

    @staticmethod
    def add_payment_document(
        db: Session,
        payment_id: int,
        file_name: str,
        mime_type: str,
        content: bytes,
        uploaded_by: int | None,
    ) -> LoanPaymentDocument:
        doc = LoanPaymentDocument(
            payment_id=payment_id,
            file_name=file_name,
            mime_type=mime_type or "application/octet-stream",
            size_bytes=len(content),
            content=content,
            uploaded_by=uploaded_by,
        )
        db.add(doc)
        db.commit()
        db.refresh(doc)
        return doc

    @staticmethod
    def get_payment_document(db: Session, doc_id: int) -> LoanPaymentDocument:
        doc = LoanDAO.get_payment_document(db, doc_id)
        if doc is None:
            raise HTTPException(status_code=404, detail="Document not found")
        return doc

    @staticmethod
    def delete_payment_document(db: Session, doc_id: int) -> None:
        doc = LoanService.get_payment_document(db, doc_id)
        db.delete(doc)
        db.commit()
